.github/workflows/haskell.yml

```
@@ -129,26 +129,45 @@ jobs:
     - name: Build all
       run: cabal build all --enable-tests
 
-    # This step is temporary. We'll soon remove it and check instead
-    # that golden files have not been modified when creating them below.
-    - name: Test all
-      run: cabal test all --enable-tests --test-show-details=direct -j1
-
     - # Delete golden files and regenerate them in the next step to ensure
       # files that tests designate as golden files are actually golden files
       # and can be generated.
       name: Delete golden files
       run: |
         rm -rf cardano-cli/test/cardano-cli-golden/files/golden
 
-    - name: Run tests regenerate golden files
+    - name: Run tests, generating golden files on the fly
       env:
         TMPDIR: ${{ runner.temp }}
         TMP: ${{ runner.temp }}
         KEEP_WORKSPACE: 1
         CREATE_GOLDEN_FILES: 1
       run: cabal test all --enable-tests --test-show-details=direct -j1
 
+    # We want this check to run first because $(git ls-files -m) (see beow) returns both
+    # modified files *and* deleted files. So we want to fail on deleted files first.
+    # This makes sure we only report modified files (and not deleted ones) in the next step,
+    # which is more intuitive.
+    - name: Check golden files are all being used
+      run: |
+        NB_UNUSED_GOLDEN_FILES=$(git ls-files -d | wc -l)
+        if [[ "$NB_UNUSED_GOLDEN_FILES" != "0" ]]; then
+          echo "⚠️ The following golden files are not used anymore:"
+          git ls-files -d
+          echo "Please delete them."
+          exit 1
+        fi
+
+    - name: Check golden files are up-to-date
+      run: |
+        NB_MODIFIED_GOLDEN_FILES=$(git ls-files -m | wc -l)
+        if [[ "$NB_MODIFIED_GOLDEN_FILES" != "0" ]]; then
+          echo "💣 The following golden files are not up-to-date:"
+          git ls-files -m
+          echo "Please run the tests locally and update the golden files, or fix your changes."
```

```suggestion
          echo "💣 The following golden files are not up-to-date:"
          echo -e
          git ls-files -m
          echo -e
          echo "These are the changes that were made:"
          echo -e
          git diff
          echo -e
          echo "Please run the tests locally and update the golden files, or fix your changes."
```
In addition to the previous suggestion, it may be useful to display the diff. That way you can see what the differences look like. So, for example, if it is a whitespace change, it may be obvious you are missing pretty-printing them. Or if a particular key in the JSON is wrong.

---

I rather keep the output of the CI short, so I will skip this one :+1: